contracts/exchange-forwarder/contracts/src/MixinExchangeWrapper.sol

```diff
@@ -223,24 +223,96 @@ contract MixinExchangeWrapper {
             totalMakerAssetAcquiredAmount = totalMakerAssetAcquiredAmount
-                .safeAdd(makerAssetAcquiredAmount);
+                .safeAdd(sellFillResults.makerAssetAcquiredAmount);
+
+            // Stop execution if the entire amount of WETH has been sold
+            if (totalWethSpentAmount >= wethSellAmount) {
+                break;
+            }
+        }
+    }
+
+    /// @dev Synchronously executes multiple calls of fillOrder until total amount of WETH (exclusive of protocol fee)
+    ///      has been sold by taker.
+    /// @param orders Array of order specifications.
+    /// @param wethSellAmount Desired amount of WETH to sell.
+    /// @param signatures Proofs that orders have been signed by makers.
+    /// @return totalWethSpentAmount Total amount of WETH spent on the given orders.
+    /// @return totalMakerAssetAcquiredAmount Total amount of maker asset acquired from the given orders.
+    function _marketSellExactAmountNoThrow(
+        LibOrder.Order[] memory orders,
+        uint256 wethSellAmount,
+        bytes[] memory signatures
+    )
+        internal
+        returns (
+            uint256 totalWethSpentAmount,
+            uint256 totalMakerAssetAcquiredAmount
+        )
+    {
+        bytes4 erc20BridgeProxyId = IAssetData(address(0)).ERC20Bridge.selector;
+        uint256 totalProtocolFeePaid;
+
+        for (uint256 i = 0; i != orders.length; i++) {
+            // Preemptively skip to avoid division by zero in _marketSellSingleOrder
+            if (orders[i].makerAssetAmount == 0 || orders[i].takerAssetAmount == 0) {
+                continue;
+            }
+
+            // The remaining amount of WETH to sell
+            uint256 remainingTakerAssetFillAmount = wethSellAmount
+                .safeSub(totalWethSpentAmount);
+
+            // If the maker asset is ERC20Bridge, take a snapshot of the Forwarder contract's balance.
+            bytes4 makerAssetProxyId = orders[i].makerAssetData.readBytes4(0);
+            address tokenAddress;
+            uint256 balanceBefore;
+            if (makerAssetProxyId == erc20BridgeProxyId) {
+                tokenAddress = orders[i].makerAssetData.readAddress(16);
+                balanceBefore = IERC20Token(tokenAddress).balanceOf(address(this));
+            }
+
+            SellFillResults memory sellFillResults = _marketSellSingleOrder(
+                orders[i],
+                signatures[i],
+                remainingTakerAssetFillAmount
+            );
+
+            // Account for the ERC20Bridge transfering more of the maker asset than expected.
+            if (makerAssetProxyId == erc20BridgeProxyId) {
+                uint256 balanceAfter = IERC20Token(tokenAddress).balanceOf(address(this));
+                sellFillResults.makerAssetAcquiredAmount = LibSafeMath.max256(
+                    balanceAfter.safeSub(balanceBefore),
+                    sellFillResults.makerAssetAcquiredAmount
+                );
+            }
```

Comment:
This all looks identical to what we do in `_marketSellNoThrow()`. Maybe we should just bake this logic into `_marketSellSingleOrder()`.